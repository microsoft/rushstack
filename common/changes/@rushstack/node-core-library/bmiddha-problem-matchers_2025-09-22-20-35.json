{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add `FileError.getProblemMatcher()` which returns the problem matcher compatible with `IOperationExecutionResult.problemCollector` as well as VS Code, GitHub Actions",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}
