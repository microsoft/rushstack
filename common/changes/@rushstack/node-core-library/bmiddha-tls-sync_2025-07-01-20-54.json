{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add a `Async.runWithTimeoutAsync` function that executes an async function, resolving if the specified timeout elapses first.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}