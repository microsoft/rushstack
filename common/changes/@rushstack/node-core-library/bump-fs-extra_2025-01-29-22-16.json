{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Update fs-extra to 11.3.0.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}