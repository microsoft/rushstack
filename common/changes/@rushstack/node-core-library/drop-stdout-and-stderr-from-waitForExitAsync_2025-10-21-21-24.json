{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Update the return type of `Executable.waitForExitAsync` to omit `stdout` and `stderr` if an `encoding` parameter isn't passed to the options object.",
      "type": "patch"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}