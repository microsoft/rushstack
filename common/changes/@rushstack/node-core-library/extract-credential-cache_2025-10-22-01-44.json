{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add \"Objects.areDeepEqual\" and \"User.getHomeFolder\" APIs.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}