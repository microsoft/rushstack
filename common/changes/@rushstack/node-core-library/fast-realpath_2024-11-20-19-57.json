{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add `RealNodeModulePathResolver` class to get equivalent behavior to `realpath` with fewer system calls (and therefore higher performance) in the typical scenario where the only symlinks in the repository are inside of `node_modules` folders and are links to package folders.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}