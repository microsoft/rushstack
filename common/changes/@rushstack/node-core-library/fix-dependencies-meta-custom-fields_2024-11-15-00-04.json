{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Extend IDependenciesMetaTable interface to support custom properties in \"dependenciesMeta\" entries",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}