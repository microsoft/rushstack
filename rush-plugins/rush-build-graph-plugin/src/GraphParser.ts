// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { Operation, IOperationRunner, ILogger } from '@rushstack/rush-sdk';

import { Colorize } from '@rushstack/terminal';
import { filterObjectForDebug } from './GraphDebugHelpers';
import { Path } from '@rushstack/node-core-library';
import type { IGraphNode } from './DropBuildGraphPlugin';

type IGraphNodeInternal = Readonly<Omit<IGraphNode, 'dependencies' | 'command'>> & {
  dependencies: ReadonlySet<string>;
  command: string | undefined;
};
type INodeMapInternal = ReadonlyMap<string, IGraphNodeInternal>;

/*
 * Try to get the operation id, return undefined if it fails
 */
export function tryGetOperationId(operation: Partial<Operation>): string | undefined {
  const task: string | undefined = operation.associatedPhase?.name;
  const project: string | undefined = operation.associatedProject?.packageName;
  return task && project ? `${project}#${task}` : undefined;
}

export class GraphParser {
  private _logger: ILogger;
  private _rushJsonFolder: string;

  public constructor(logger: ILogger, rushJsonFolder: string) {
    this._logger = logger;
    this._rushJsonFolder = Path.convertToSlashes(rushJsonFolder);
    if (!this._rushJsonFolder.endsWith('/')) {
      this._rushJsonFolder += '/';
    }
  }

  /*
   * Convert the operationMap into an array of graph nodes with empty commands pruned
   */
  public processOperations(operations: Iterable<Operation>): IGraphNode[] {
    const result: Map<string, IGraphNodeInternal> = new Map();
    for (const operation of operations) {
      const entry: IGraphNodeInternal = this._operationAsHashedEntry(operation);
      result.set(entry.id, entry);
    }

    return this._pruneNoOps(result);
  }

  /*
   * Validate that all dependencies exist
   * Validate that all nodes have a non-empty command
   * Print a message to the logger, and return true if the graph is valid
   */
  public validateGraph(entries: Iterable<Readonly<IGraphNode>>): boolean {
    const entryIDs: Set<string> = new Set();
    for (const entry of entries) {
      entryIDs.add(entry.id);
    }
    const { terminal } = this._logger;
    let totalEdges: number = 0;
    let totalNodes: number = 0;
    let isValid: boolean = true;
    for (const entry of entries) {
      for (const depId of entry.dependencies) {
        if (!entryIDs.has(depId)) {
          terminal.writeErrorLine(`${entry.id} has a dependency on ${depId} which does not exist`);
          isValid = false;
        }
      }

      totalEdges += entry.dependencies.length;
      totalNodes++;

      if (!entry.id) {
        terminal.writeErrorLine(`An entry id is missing or empty`);
        isValid = false;
      }

      if (!entry.command) {
        terminal.writeErrorLine(`There is an empty command in ${entry.id} `);
        isValid = false;
      }
    }

    if (isValid) {
      terminal.writeLine(Colorize.green('All nodes have non-empty commands and dependencies which exist'));
    }
    terminal.writeLine(`Graph has ${totalNodes} nodes, ${totalEdges} edges`);
    return isValid;
  }

  /*
   * Get the operation id, throw an error if it fails
   */
  public getOperationId(operation: Operation): string {
    const result: string | undefined = tryGetOperationId(operation);
    if (!result) {
      throw new Error(
        'Operation does not have a name: ' + JSON.stringify(filterObjectForDebug(operation, 2), undefined, 2)
      );
    }
    return result;
  }

  /*
   * remove all entries with empty commands
   * if an entry has a dependency with an empty command, it should inherit the dependencies of the empty command
   */
  private _pruneNoOps(inputNodeMap: INodeMapInternal): IGraphNode[] {
    // Cache for the non-empty upstream dependencies of each operation
    const nonEmptyDependenciesByOperation: Map<IGraphNodeInternal, Set<string>> = new Map();
    function getNonEmptyDependencies(node: IGraphNodeInternal): ReadonlySet<string> {
      // If we've already computed this, return the cached result
      let nonEmptyDependencies: Set<string> | undefined = nonEmptyDependenciesByOperation.get(node);
      if (!nonEmptyDependencies) {
        nonEmptyDependencies = new Set();
        nonEmptyDependenciesByOperation.set(node, nonEmptyDependencies);
        for (const dependencyID of node.dependencies) {
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          if (!inputNodeMap.get(dependencyID)!.command) {
            // If the dependency is empty, recursively inherit its non-empty dependencies
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            for (const deepDependency of getNonEmptyDependencies(inputNodeMap.get(dependencyID)!)) {
              nonEmptyDependencies.add(deepDependency);
            }
          } else {
            nonEmptyDependencies.add(dependencyID);
          }
        }
      }
      return nonEmptyDependencies;
    }

    const result: IGraphNode[] = [];
    for (const node of inputNodeMap.values()) {
      if (node.command) {
        result.push({
          ...node,
          dependencies: Array.from(getNonEmptyDependencies(node)),
          command: node.command
        });
      }
    }

    return result;
  }

  /*
   * Convert an operation into a graph node
   */
  private _operationAsHashedEntry(operation: Operation): IGraphNodeInternal {
    const dependencies: Set<string> = new Set();
    for (const dep of operation.dependencies.values()) {
      dependencies.add(this.getOperationId(dep));
    }

    let workingDirectory: string | undefined = operation.associatedProject?.projectFolder;
    if (workingDirectory && workingDirectory.startsWith(this._rushJsonFolder)) {
      workingDirectory = workingDirectory.replace(this._rushJsonFolder, '');
    }

    const node: Partial<IGraphNodeInternal> = {
      id: tryGetOperationId(operation),
      task: operation.associatedPhase?.name,
      package: operation.associatedProject?.packageName,
      dependencies,
      workingDirectory,
      // _commandToRun does not exist in the type definition of IOperationRunner,
      //    but it's defined in subclasss ShellOperationRunner
      command: (operation.runner as IOperationRunner & { _commandToRun: string })?._commandToRun
    };

    // command is absent because it is not required
    const requiredFields: Array<keyof IGraphNodeInternal> = [
      'id',
      'task',
      'package',
      'dependencies',
      'workingDirectory'
    ];
    const missingFields: Array<keyof IGraphNodeInternal> = requiredFields.filter((prop) => !node[prop]);

    if (missingFields.length > 0) {
      this._logger.terminal.writeErrorLine(
        `Operation is missing required fields ${missingFields.join(', ')}: ${JSON.stringify(node)}`
      );
    }

    // the runner is a no-op if and only if the command is empty
    if (!!operation.runner?.isNoOp !== !node.command) {
      throw new Error(`${node.id}: Operation runner isNoOp does not match _commandToRun existence`);
    }

    return node as IGraphNodeInternal;
  }
}
