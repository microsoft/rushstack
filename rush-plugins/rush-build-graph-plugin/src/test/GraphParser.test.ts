// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { JsonFile, PackageJsonLookup } from '@rushstack/node-core-library';
import type { ILogger, Operation } from '@rushstack/rush-sdk';
import { RushConfiguration } from '@rushstack/rush-sdk';

import type { IGraphNode } from '../DropBuildGraphPlugin';
import { GraphParser } from '../GraphParser';

import testData from './test-operation-map.json';

const lookup: PackageJsonLookup = new PackageJsonLookup();
lookup.tryGetPackageFolderFor(__dirname);

const thisProjectFolder: string | undefined = lookup.tryGetPackageFolderFor(__dirname);
if (!thisProjectFolder) {
  throw new Error('Cannot find project folder');
}

const rushJsonFolder: string | undefined = RushConfiguration.tryFindRushJsonLocation();
if (!rushJsonFolder) {
  throw new Error('Cannot find rush.json folder');
}

// in order to update this file, run the following in the rushBuildGraphPluginTestRepo folder:
// rush test --drop-graph ../examples/graph.json
const graphLocation: string = `${thisProjectFolder}/examples/graph.json`;
const graph = JsonFile.load(graphLocation);

const exampleGraph: readonly IGraphNode[] = Array.from(graph.nodes as IGraphNode[]).sort((a, b) =>
  a.id > b.id ? 1 : -1
) as readonly IGraphNode[];

const graphParser: GraphParser = new GraphParser(
  {
    terminal: { writeErrorLine: jest.fn(), writeLine: jest.fn() }
  } as unknown as ILogger,
  rushJsonFolder
);

// These run unit tests on the checked in test-operation-map.json file, which mimics what rush passes in
describe(GraphParser.name, () => {
  it('should fail if the input schema is invalid', () => {
    const operations = JSON.parse(JSON.stringify(testData.OperationMap));
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (operations[0].dependencies as Array<any>).push({
      incorrectPhase: { name: 'incorrectPhase' },
      incorrectProject: { packageName: 'incorrectProject' }
    });
    expect(() => graphParser.processOperations(new Set(operations))).toThrowErrorMatchingSnapshot();
  });

  it('should fail if isNoOp mismatches a command', () => {
    const operations = JSON.parse(JSON.stringify(testData.OperationMap));
    const firstOperation: Operation = operations[0];
    // @ts-ignore, isNoOp is a readonly property
    firstOperation.runner!.isNoOp = true;
    // @ts-ignore, _commandToRun doesn't exist on IOperationRunner
    firstOperation.runner!._commandToRun = 'echo "hello world"';
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(() => graphParser.processOperations(new Set(operations as any))).toThrowErrorMatchingSnapshot();
  });
});

// These check the graph.json file that was generated by the integration test and unit tests the validator on it
describe('validateGraph', () => {
  it('should validate graph.json', () => {
    const isValid: boolean = graphParser.validateGraph(exampleGraph);
    expect(isValid).toBe(true);
  });
  it('should fail to validate when command is empty', () => {
    const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
    invalidGraph[1].command = '';
    const isValid: boolean = graphParser.validateGraph(invalidGraph);
    expect(isValid).toBe(false);
  });
  it('should fail to validate when command is missing', () => {
    const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
    delete invalidGraph[1].command;
    const isValid: boolean = graphParser.validateGraph(invalidGraph);
    expect(isValid).toBe(false);
  });
  it('should fail to validate when id is missing', () => {
    const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
    delete invalidGraph[1].id;
    const isValid: boolean = graphParser.validateGraph(invalidGraph);
    expect(isValid).toBe(false);
  });
  it('should fail to validate when id is empty', () => {
    const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
    invalidGraph[1].id = '';
    const isValid: boolean = graphParser.validateGraph(invalidGraph);
    expect(isValid).toBe(false);
  });
  it('should fail to validate when dependencies refer to nonexistent nodes', () => {
    const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
    invalidGraph[1].dependencies.push('@this/is/presumably#_not:real');
    const isValid: boolean = graphParser.validateGraph(invalidGraph);
    expect(isValid).toBe(false);
  });
});
