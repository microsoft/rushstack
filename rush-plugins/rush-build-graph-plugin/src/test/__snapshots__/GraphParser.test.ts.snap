// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--drop-graph integration process Should produce consistent output graph 1`] = `
Array [
  Object {
    "command": "heft run build command ",
    "dependencies": Array [],
    "id": "package-a#_phase:build",
    "package": "package-a",
    "task": "_phase:build",
    "workingDirectory": "a",
  },
  Object {
    "command": "heft run a different build command ",
    "dependencies": Array [
      "package-a#_phase:build",
    ],
    "id": "package-b#_phase:build",
    "package": "package-b",
    "task": "_phase:build",
    "workingDirectory": "b",
  },
  Object {
    "command": "heft run a test command ",
    "dependencies": Array [
      "package-b#_phase:build",
    ],
    "id": "package-b#_phase:test",
    "package": "package-b",
    "task": "_phase:test",
    "workingDirectory": "b",
  },
]
`;

exports[`GraphParser should fail if isNoOp mismatches a command 1`] = `"a#_phase:build: Operation runner isNoOp does not match _commandToRun existence"`;

exports[`GraphParser should fail if the input schema is invalid 1`] = `
"Operation does not have a name: {
  \\"incorrectPhase\\": {
    \\"name\\": \\"incorrectPhase\\"
  },
  \\"incorrectProject\\": {
    \\"packageName\\": \\"incorrectProject\\"
  }
}"
`;
